metricbeat/module/windows/perfmon/pdh_windows.go

```diff
@@ -1,133 +1,237 @@
+	if err != nil {
+		return errors.Wrapf(err, `failed to add counter (path="%v")`, counterPath)
+	}
 
-	err := _PdhCollectQueryData(q.query)
+	q.counters[counterPath] = h
+	return nil
+}
+
+func (q *Query) Execute() error {
+	return PdhCollectQueryData(q.handle)
+}
 
-	if firstFetch {
-		// Most counters require two sample values in order to compute a displayable value. So wait and then collect the second value
-		time.Sleep(2000)
-		err = _PdhCollectQueryData(q.query)
+type Value struct {
+	Num float64
+	Err error
+}
+
+func (q *Query) Values() (map[string]Value, error) {
+	rtn := make(map[string]Value, len(q.counters))
+	for path, handle := range q.counters {
+		_, value, err := PdhGetFormattedCounterValue(handle, PdhFmtDouble|PdhFmtNoCap100)
+		if err != nil {
+			rtn[path] = Value{Err: err}
+			continue
+		}
+
+		rtn[path] = Value{Num: *(*float64)(unsafe.Pointer(&value.LongValue))}
 	}
 
-	if err != ERROR_SUCCESS {
-		return nil, PdhError(err)
+	return rtn, nil
+}
+
+// Closes the query and all of its counters.
+func (q *Query) Close() error {
+	return PdhCloseQuery(q.handle)
+}
+
+type PerfmonReader struct {
+	query     *Query            // PDH Query
+	pathToKey map[string]string // Mapping of counter path to key used in output.
+	executed  bool              // Indicates if the query has been executed.
+}
+
+func NewPerfmonReader(config []CounterConfig) (*PerfmonReader, error) {
+	query, err := NewQuery("")
+	if err != nil {
+		return nil, err
 	}
 
-	result := common.MapStr{}
+	r := &PerfmonReader{
+		query:     query,
+		pathToKey: map[string]string{},
+	}
 
-	for _, v := range q.counters {
-		groupVal := make(map[string]interface{})
-		for _, v1 := range v.Counters {
-			err := _PdhGetFormattedCounterValue(v1.counter, PdhFmtDouble, q.counterType, &v1.displayValue)
-			if err != ERROR_SUCCESS {
-				switch err {
-				case PDH_CALC_NEGATIVE_DENOMINATOR:
-					{
-						//Sometimes counters return negative values. We can ignore this error. See here for a good explanation https://www.netiq.com/support/kb/doc.php?id=7010545
-						groupVal[v1.counterName] = 0
-						continue
-					}
-				default:
-					{
-						return nil, PdhError(err)
-					}
-				}
+	for _, group := range config {
+		for _, collector := range group.Group {
+			if err := query.AddCounter(collector.Query); err != nil {
+				query.Close()
+				return nil, err
 			}
-			doubleValue := (*float64)(unsafe.Pointer(&v1.displayValue.LongValue))
-			groupVal[v1.counterName] = *doubleValue
 
+			r.pathToKey[collector.Query] = group.Name + "." + collector.Alias
 		}
-		result[v.GroupName] = groupVal
 	}
-	return result, 0
+
+	return r, nil
 }
 
-func CloseQuery(q uintptr) PdhError {
-	err := _PdhCloseQuery(q)
-	if err != ERROR_SUCCESS {
-		return PdhError(err)
+func (r *PerfmonReader) Read() (common.MapStr, error) {
+	if !r.executed {
+		// Most counters require two sample values in order to compute a
+		// displayable value. So initial perform two samples.
+		r.query.Execute()
```

Comment:
Hi @andrewkroh. You wrote 
>but the first value will always be 0

This is because the time slot is to small between these two calls. I would prefer the following solution
```golang
if err := r.query.Execute(); err != nil {
		return nil, err
	}

	// Get the values.
	values, err := r.query.Values()
	if err != nil {
		return nil, errors.Wrap(err, "failed formatting counter values")
	}

	// Write the values into the map.
	result := common.MapStr{}
	var errs multierror.Errors

	for counterPath, value := range values {
		if value.Err != nil {
			// XXX: Sometimes counters return negative values. We can ignore
			// this error. For an explanation see:
			// https://www.netiq.com/support/kb/doc.php?id=7010545
			if value.Err == PDH_CALC_NEGATIVE_DENOMINATOR {
				result.Put(counterPath, 0)
				continue
			}
			if value.Err == PDH_INVALID_DATA {
				if !r.executed {
					// Most counters require two sample values in order to compute a
					// displayable value. So the first fetch for this counters return PDH_INVALID_DATA.
					// So check if this is the first fetch. If so just move on and dont't put that 0 value
					// into our result. If not this is maybe a real error
					r.executed = true
					continue
				}
			}
			errs = append(errs, value.Err)
			continue
```